### DISCUSSION TITLE
0608 피어세션일지

### DISCUSSION THREAD

**Discussion by gvsteve24** [bcaitech1/p4-dkt-freshtomato#97] [General]
# 오늘의 TMI + 이모지 

| 재훈  | 민용    | 유라         | 아라      | 정현              | 진현        |
| ----- | ------- | ------------ | --------- | ----------------- | ----------- |
| :pig: | :bento: | :woozy_face: | :pensive: | :man_facepalming: | :happy2sad: |

# 회의 및 토론

### 실험 진행 상황 

[유라] [Discussion](https://github.com/bcaitech1/p4-dkt-freshtomato/discussions/81)에 Custom KFold 추가 작성 (sklearn source code k-fold validation 검색 _ [스택오버플로](https://stackoverflow.com/questions/30040597/how-to-generate-a-custom-cross-validation-generator-in-scikit-learn/34195078))

10fold결과가 일정하지 않음 < 데이터가 작아서 그런 것 같다.

[재훈] [Discussion](https://github.com/bcaitech1/p4-dkt-freshtomato/discussions/89)에 Riiid 데이터 공유 (전처리를 해서 좀 맞춰줌)

* tag 클러스터링을 못한게 아쉬움

* contentId , task_container_id로 assessmentItemID 컬럼과 비슷하게 구성 ( 9454개로 가져옴 )

  아라) riiid 데이터랑 scream 데이터랑 수치가 겹치면  pretraining에 문제가 생길거같다.

  민용) pretrain하고 embedding layer는 못가져온다고 했을때 
  => backbone의 transformer만 가지고 올때 
  => feature로 넣어주는 값들이 문제가 되는 경우가 있을거같다

  

Pretrain한 모델을 이어서 학습하는 측면에서 새로운 Embedding layer를 쓰기 땜에 문제가 없을거 같다. ( 이유가. ) 

정현) pretrain이라는게 데이터를 그대로 가져가고 학습된 weight를 이용하는거 아닌가?

민용) 데이터를 같은 분포를 가지도록 만졌더라도 의미가 다르기 때문에 pretrain의 의미가 퇴색되지는 않을것 같다.
민용) 데이터 감사! 학습시키고 유효한지 평가하는 건 DL팀이 하면 된다.
민용) 데이터를 수정, 정제해준 부분이 좋아보임. 원래 의미가 아니게 되는건 확인해봐야 할 부분

아라) 도출한 데이터에서 성능이 안나오는 거는 riiid데이터 전처리 과정에서 고려하지 않은 부분이 있었을 수 있다. 
아라) 기존의 캐글을 봐도 성능이 나오는 경우가 존재하니까 이미 성능이 나오는 예에서 가져와서 볼 수도 있을 것 같다.

민용) 실험을 해봐야겠다. 유의미한 산출물임.

10 fold로 20 epoch에 3시간 40분 정도 소요되는데, pretrain은 한번만 돌려도 될 것 같다. 

last query를 pretrain돌리는게 어떨지!? Transformer같은 것들은 다 pretrain하면 좋을거같음. 

모델!  각자 나눠서 돌려보고 결과 확인하기

모델을 정한걸로 pretrain하고.. pretrain된거 학습시켜서 또 확인하기

embedding layer도 pretrain/ no pretrain 해보기

아라) 성적...이 나와야 쓸텐데..... 지금 63점 나온걸로는... 남은 20을 올리는데에 살짝 회의적이다.

민용) 유의미한 피처들을 넣어준 상태로 또 실험해보자 

**Last Query 실험 상황**

Valid AUC: 0.8074 LB: 0.7745 (차이: 0.0329)

0 Fold Best Valid AUC:0.8263058436593915
1 Fold Best Valid AUC:0.8112963286204689
2 Fold Best Valid AUC:0.8283297791592872
3 Fold Best Valid AUC:0.7967976289442295
4 Fold Best Valid AUC:0.8186662617586025
5 Fold Best Valid AUC:0.7959438851104735
6 Fold Best Valid AUC:0.8141720161501146
7 Fold Best Valid AUC:0.7907464153561516
8 Fold Best Valid AUC:0.8092715327517543
9 Fold Best Valid AUC:0.7825186922314146
Average AUC:0.807404838374189

[민용]

time_elapsed, user_cumcount, time_elapsed_cat , 300 cut

test_id_mean 을 짤라서 넣어봤는데 별로 좋진 않았다.

재훈) last kfold 돌릴때 훈련 오래걸렸어?
유라) 엄청 오래걸렸음아마 기존에 정의된 것들은 속도를 좀 최적화하게 짜여 있나봄, 표 뜰때까지 거의 40분 걸리고 finalize 할 때도 그정도 걸림

[진현] 피처 추가 ([문제난이도 이론](https://m.blog.naver.com/PostView.naver?isHttpsRedirect=true&blogId=paean0330&logNo=40118009160))

```python
whole_count = df.groupby('assessmentItemID').size()
correct_count = df.groupby('assessmentItemID').answerCode.sum()
wrong_count =  whole_count - correct_count
df['time_elapsed_over_300'] = 0
df.loc[list(df[df.time_elapsed > 300].index), 'time_elapsed_over_300'] = 1
over_count = df.groupby('assessmentItemID').time_elapsed_over_300.sum()

Level_assessment = (correct_count - wrong_count) / (whole_count - over_count) * 100

```

너무 오래걸리면 빼주고(300초 컷) 
정답자에서 오답자를 빼주는 코드 => 문제자체의 난이도

20~70%여야되는데 음수가 나옴 (오답자가 더 많은) 

* 300초 이상 인 경우 100 % 넘어가는 것도 있다.

### 향후 실험 계획

아래에 공유부탁합니다:)

# 저녁메뉴 공유공유

| 재훈 | 민용    | 유라          | 아라          | 정현       | 진현          |
| ---- | ------- | ------------- | ------------- | ---------- | ------------- |
| ❓ | :bento: | :green_salad: | :green_salad: | :dumpling: | :green_salad: |

**Comment by arabae** [comment]
**_TODO_**

- [x] LastQuery tunning
- [x] Riiid data 사용하여 train 수행

**Comment by gvsteve24** [comment]
- [ ] LTSMATT 정리
- [ ] LSTM실험 전체데이터
- [ ] LSTMATT실험 전체데이터
- [ ] attention개념 정리

**Comment by yukyunglee** [comment]
남은 대회 기간동안은 저도 discuss를 보고 같이 코멘트를 다는게 좋겠군요 ! 하나 질문드리고자 하는것은 여기서 말씀하시는 pretrain은 정확히 어떤것을 의미하나요 ?

**Reply by arabae** [reply]
멘토님 코멘트 감사합니다 :) 올려주신 멘토링 정리에 답변을 달아두었습니다.
